File: backend/services/phone.py
"""National phone length by country calling code. Keep in sync with frontend/src/lib/phone.ts."""

from __future__ import annotations

import re

# iso, name, dial (no plus), national digit length min/max
PHONE_COUNTRIES: tuple[tuple[str, str, str, int, int], ...] = (
    ("IN", "India", "91", 10, 10),
    ("US", "United States", "1", 10, 10),
    ("GB", "United Kingdom", "44", 10, 10),
    ("AE", "United Arab Emirates", "971", 9, 9),
    ("SG", "Singapore", "65", 8, 8),
    ("AU", "Australia", "61", 9, 9),
    ("DE", "Germany", "49", 10, 11),
    ("FR", "France", "33", 9, 9),
    ("SA", "Saudi Arabia", "966", 9, 9),
    ("PK", "Pakistan", "92", 10, 10),
    ("BD", "Bangladesh", "880", 10, 10),
    ("NP", "Nepal", "977", 10, 10),
    ("LK", "Sri Lanka", "94", 9, 9),
    ("MY", "Malaysia", "60", 9, 10),
    ("PH", "Philippines", "63", 10, 10),
    ("ID", "Indonesia", "62", 9, 11),
    ("NL", "Netherlands", "31", 9, 9),
    ("ZA", "South Africa", "27", 9, 9),
    ("CA", "Canada", "1", 10, 10),
)

_BY_DIAL: dict[str, list[tuple[str, str, int, int]]] = {}
for iso, name, dial, lo, hi in PHONE_COUNTRIES:
    _BY_DIAL.setdefault(dial, []).append((iso, name, lo, hi))

_DIGITS = re.compile(r"\D+")


def _digits(value: str) -> str:
    return _DIGITS.sub("", value or "")


def validate_e164_phone(raw: str) -> str:
    """Return +<dial><national> or raise ValueError with a user-facing sentence."""
    compact = (raw or "").strip().replace(" ", "")
    if not compact:
        raise ValueError("Enter your phone number.")

    if not compact.startswith("+"):
        raise ValueError("Select a country code and enter your phone number.")

    digits = _digits(compact)
    if not digits:
        raise ValueError("Enter digits only.")

    # Longest dial match (1, 44, 91, 971, …)
    match: tuple[str, str, int, int] | None = None
    matched_dial = ""
    for size in (3, 2, 1):
        prefix = digits[:size]
        options = _BY_DIAL.get(prefix)
        if options:
            match = options[0]
            matched_dial = prefix
            break
    if match is None:
        raise ValueError("Select a valid country code.")

    _iso, name, lo, hi = match
    national = digits[len(matched_dial) :]
    if national.startswith("0"):
        national = national.lstrip("0")
    if not national.isdigit() or not national:
        raise ValueError("Enter digits only.")
    n = len(national)
    if n < lo or n > hi:
        if lo == hi:
            raise ValueError(f"{name} numbers must be {lo} digits.")
        raise ValueError(f"{name} numbers must be {lo}–{hi} digits.")
    return f"+{matched_dial}{national}"

File: frontend/src/lib/phone.ts
/** Keep in sync with backend/services/phone.py */

export interface PhoneCountry {
  iso: string
  name: string
  dial: string
  min: number
  max: number
}

export const PHONE_COUNTRIES: PhoneCountry[] = [
  { iso: 'IN', name: 'India', dial: '91', min: 10, max: 10 },
  { iso: 'US', name: 'United States', dial: '1', min: 10, max: 10 },
  { iso: 'GB', name: 'United Kingdom', dial: '44', min: 10, max: 10 },
  { iso: 'AE', name: 'United Arab Emirates', dial: '971', min: 9, max: 9 },
  { iso: 'SG', name: 'Singapore', dial: '65', min: 8, max: 8 },
  { iso: 'AU', name: 'Australia', dial: '61', min: 9, max: 9 },
  { iso: 'DE', name: 'Germany', dial: '49', min: 10, max: 11 },
  { iso: 'FR', name: 'France', dial: '33', min: 9, max: 9 },
  { iso: 'SA', name: 'Saudi Arabia', dial: '966', min: 9, max: 9 },
  { iso: 'PK', name: 'Pakistan', dial: '92', min: 10, max: 10 },
  { iso: 'BD', name: 'Bangladesh', dial: '880', min: 10, max: 10 },
  { iso: 'NP', name: 'Nepal', dial: '977', min: 10, max: 10 },
  { iso: 'LK', name: 'Sri Lanka', dial: '94', min: 9, max: 9 },
  { iso: 'MY', name: 'Malaysia', dial: '60', min: 9, max: 10 },
  { iso: 'PH', name: 'Philippines', dial: '63', min: 10, max: 10 },
  { iso: 'ID', name: 'Indonesia', dial: '62', min: 9, max: 11 },
  { iso: 'NL', name: 'Netherlands', dial: '31', min: 9, max: 9 },
  { iso: 'ZA', name: 'South Africa', dial: '27', min: 9, max: 9 },
  { iso: 'CA', name: 'Canada', dial: '1', min: 10, max: 10 },
]

export const DEFAULT_PHONE_ISO = 'IN'

export function phoneCountry(iso: string): PhoneCountry {
  return PHONE_COUNTRIES.find((row) => row.iso === iso) ?? PHONE_COUNTRIES[0]
}

export function nationalDigits(value: string): string {
  return value.replace(/\D/g, '')
}

export function validateNationalNumber(iso: string, raw: string): { ok: true; e164: string } | { ok: false; message: string } {
  const country = phoneCountry(iso)
  let national = nationalDigits(raw)
  if (national.startsWith('0')) national = national.replace(/^0+/, '')
  if (!national) {
    return { ok: false, message: 'Enter your phone number.' }
  }
  if (national.length < country.min || national.length > country.max) {
    if (country.min === country.max) {
      return { ok: false, message: `${country.name} numbers must be ${country.min} digits.` }
    }
    return {
      ok: false,
      message: `${country.name} numbers must be ${country.min}–${country.max} digits.`,
    }
  }
  return { ok: true, e164: `+${country.dial}${national}` }
}

const PHONE_EXAMPLES: Record<string, string> = {
  IN: '98765 43210',
  US: '202 555 0100',
  GB: '7400 123456',
  AE: '50 123 4567',
  SG: '8123 4567',
  AU: '412 345 678',
  DE: '151 2345678',
  FR: '6 12 34 56 78',
  SA: '51 234 5678',
  PK: '300 1234567',
  BD: '1712 345678',
  NP: '9841 234567',
  LK: '71 234 5678',
  MY: '12 345 6789',
  PH: '917 123 4567',
  ID: '812 3456 789',
  NL: '6 12345678',
  ZA: '82 123 4567',
  CA: '416 555 0100',
}

export function phonePlaceholder(iso: string): string {
  return PHONE_EXAMPLES[iso] ?? 'Your number'
}

export function countryFlagSrc(iso: string): string {
  return `https://flagcdn.com/w40/${iso.toLowerCase()}.png`
}
